modules/core/src/test/java/org/apache/ignite/internal/processors/continuous/IgniteContinuousQueryMetadataUpdateTest.java

```diff
@@ -57,11 +55,26 @@
     @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
         IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
 
-        TestDiscoverySpi discoSpi = new TestDiscoverySpi();
+        TestTcpDiscoverySpi discoSpi = (TestTcpDiscoverySpi)cfg.getDiscoverySpi();
 
-        discoSpi.setIpFinder(IP_FINDER);
+        DiscoveryHook discoveryHook = new DiscoveryHook() {
+            @Override public void beforeDiscovery(DiscoverySpiCustomMessage msg) {
+                DiscoveryCustomMessage customMsg = msg == null ?
```

Comment:
I think we should provide `customMsg` as argument  (or instead of `DiscoverySpiCustomMessage msg`) and do not duplicate:
```
DiscoveryCustomMessage customMsg = msg == null ?
    null : (DiscoveryCustomMessage)IgniteUtils.field(msg, "delegate");
```